packages/react-components/react-tree/src/components/TreeItem/useTreeItem.ts

```diff
@@ -0,0 +1,109 @@
+  const { 'aria-owns': ariaOwns, as = 'div', onKeyDown, ...rest } = props;
+
+  const { openTrees, requestOpenChange, level } = useTreeContext_unstable();
+  const { findFirstFocusable } = useFocusFinders();
+  const { targetDocument } = useFluent_unstable();
+
+  const triggerRef = useTriggerRef();
+
+  const isBranch = typeof ariaOwns === 'string';
+  const isOpen = React.useMemo(() => isBranch && openTrees.includes(ariaOwns!), [ariaOwns, openTrees, isBranch]);
+
+  const handleClick = useEventCallback((event: React.MouseEvent<TreeItemElementIntersection>) => {
+    if (isBranch) {
+      requestOpenChange({ event, open: !isOpen, type: 'click', id: ariaOwns! });
+    }
+  });
+  const handleArrowRight = (event: React.KeyboardEvent<TreeItemElementIntersection>) => {
+    if (isOpen && isBranch) {
+      // find first focusable on the subtree and focus on it
+      const subtree = targetDocument?.getElementById(ariaOwns!);
+      if (subtree) {
+        subtree && findFirstFocusable(subtree)?.focus();
+      }
+    }
+    if (isBranch && !isOpen) {
+      requestOpenChange({ event, open: true, type: 'arrowRight', id: ariaOwns! });
+    }
+  };
+  const handleArrowLeft = (event: React.KeyboardEvent<TreeItemElementIntersection>) => {
+    if (!isBranch || !isOpen) {
+      triggerRef.current?.focus();
+    }
+    if (isBranch && isOpen) {
+      requestOpenChange({ event, open: false, type: 'arrowLeft', id: ariaOwns! });
+    }
+  };
+  const handleKeyDown = useEventCallback((event: React.KeyboardEvent<TreeItemElementIntersection>) => {
+    onKeyDown?.(event);
+    if (event.isDefaultPrevented()) {
+      return;
+    }
+    switch (event.code) {
+      case ArrowRight: {
+        return handleArrowRight(event);
+      }
+      case ArrowLeft: {
+        return handleArrowLeft(event);
+      }
+    }
+  });
+  return {
+    components: {
+      root: 'div',
+    },
+    isOpen,
+    isLeaf: !isBranch,
+    root: getNativeElementProps(
+      as,
+      useARIAButtonProps(as, {
+        ...rest,
+        // casting here is required to convert union to intersection
+        ref: ref as React.Ref<TreeItemElementIntersection>,
+        'aria-owns': ariaOwns,
+        'aria-level': level,
+        // FIXME: tabster fails to navigate when aria-expanded is true
+        // 'aria-expanded': isBranch ? isOpen : undefined,
+        role: 'treeitem',
+        onClick: handleClick,
+        onKeyDown: handleKeyDown,
+      }),
+    ),
+  };
+};
+
+/**
+ * retrieves a reference to the treeitem that triggered an opening of a subtree
+ */
+function useTriggerRef() {
+  const { subtreeRef, treeRef } = useTreeContext_unstable();
+  const triggerRef = React.useRef<TreeItemElementIntersection | null>(null);
+
+  React.useEffect(() => {
+    if (subtreeRef.current) {
+      triggerRef.current = treeRef.current?.querySelector(`[aria-owns="${subtreeRef.current.id}"]`) ?? null;
```

Comment:
I don't think querySelector from root is necessary here, you should just be able to create a tree walker and walk in reverse order, you should be guaranteed to find the parent

Reply:
If I do a tree walker and navigate upper to find the previous sibling of the parent element, I'm assuming so much of the DOM structure,  the user can anytime wrap it's own structure around each item, breaking this assumptions.

I'm not sure how reliable traversing it would be.

Reply:
even if the user wraps around the item, the walker would still find the correct node with `aria-owns`. The end result would be that the tree walk would end up on the root tree anyway. However, it should always stop at the sibling of its immediate parent, what would be the case that the walker wouldn't find something with `aria-owns`?

Reply:
anyway I won't block on this, it might not be a big deal 😀

Reply:
Some crazy things like these are possible:

```tsx
    <Tree aria-label="Tree">
      <TreeItem aria-owns="default-subtree-1">level 1, item 1</TreeItem>
      <TreeItem aria-owns="default-subtree-2">level 1, item 2</TreeItem>
      <Tree id="default-subtree-2">
        <TreeItem aria-owns="default-subtree-2-1">level 2, item 1</TreeItem>
        <Tree id="default-subtree-2-1">
          <TreeItem>level 3, item 1</TreeItem>
        </Tree>
      </Tree>
      <Tree id="default-subtree-1">
        <TreeItem>level 2, item 1</TreeItem>
        <TreeItem>level 2, item 2</TreeItem>
        <TreeItem>level 2, item 3</TreeItem>
      </Tree>
    </Tree>
```

Reply:
another solution I was thinking was to have a `Map<string, React.Ref>` of `"owners"`?! WDYT? this would avoid the query, but it could get bigger... if there are a bunch of `"owners"`